File: 140. Word Break II/C++/Solution.cpp
class Solution {
public:
    vector<string> wordBreak(string s, vector<string> &wordDict) {
        class Utils {
        public:
            vector<string>
            backTrace(string s, vector<string> &wordDict, unordered_map<string, vector<string>> &memory) {
                if (s.empty()) {
                    vector<string> result;
                    result.push_back("");
                    return result;
                }
                if (memory.find(s) != memory.end()) {
                    return memory[s];
                }
                vector<string> currentMemory;
                for (auto &word : wordDict) {
                    if (s.find(word) == 0) {
                        vector<string> lefts = backTrace(s.substr(word.length()), wordDict, memory);
                        for (auto &left : lefts) {
                            currentMemory.push_back(word + (left.empty() ? "" : " ") + left);
                        }
                    }
                }
                memory[s] = currentMemory;
                return currentMemory;
            }
        };
        Utils utils;
        unordered_map<string, vector<string>> temp;
        return utils.backTrace(s, wordDict, temp);
    }
};

File: 140. Word Break II/Python/Solution.py
class Solution(object):
    def wordBreak(self, s, wordDict):
        """
        :type s: str
        :type wordDict: List[str]
        :rtype: List[str]
        """
        def back_trace(s, word_dict, current, result, memory):
            if not s:
                result.append(' '.join(current))
                return ['']
            if s in memory:
                if memory[s]:
                    part = ' '.join(current)
                    for item in memory[s]:
                        result.append(part + ' ' + item)
                return memory[s]
            else:
                current_memory = []
                for word in word_dict:
                    if s.startswith(word):
                        current.append(word)
                        lefts = back_trace(s[len(word):], word_dict, current, result, memory)
                        current.pop()
                        current_memory.extend(
                            [word + (' ' if left else '') + left for left in lefts])
                memory[s] = current_memory
                return current_memory

        result = []
        back_trace(s, set(wordDict), [], result, {})
        return result
